feat: S3 fallback for event capture

frankh opened the pull request:
## Problem

We currently have a major single point of failure for event capture: Kafka

If Kafka (or warpstream) goes down we start losing 100% of new captured events. Implement a fallback where events are sent to S3 if kafka fails, so we can reingest them once kafka is back up

## Changes

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?

<!-- Yes / no / it doesn't have an impact. -->

## How did you test this code?

ran on 100% production traffic mirrored to warpstream, shut down warpstream agents and observed it fallback to s3

frankh committed:
Add S3Sink and FallbackSink to rust capture

This allows us to have events sent to S3 as a fallback if the KafkaSink fails,
removing kafka as a single point of failure

benjackwhite commented on rust/capture/src/config.rs:
```diff
@@ -61,6 +61,55 @@ pub struct Config {
+            overflow_burst_limit: NonZeroU32::new(1000).unwrap(),
+            overflow_forced_keys: None,
+            dropped_keys: None,
+            kafka: KafkaConfig {
```
Isn't this a duplicate of the stuff below 🤔

frankh replied:
Yeah this needs a refactor tbh

benjackwhite commented on rust/capture/src/server.rs:
```diff
@@ -120,26 +122,57 @@ where
+                .await;
+
+            let s3_sink = S3Sink::new(
+                config
+                    .s3_fallback_bucket
+                    .expect("S3 bucket required when fallback enabled"),
+                config.s3_fallback_endpoint,
+                sink_liveness,
+            )
+            .await
+            .expect("failed to create S3 sink");
+
+            let fallback_sink = FallbackSink::new(kafka_sink, s3_sink);
+
+            router::router(
+                crate::time::SystemTime {},
+                liveness,
+                fallback_sink,
+                redis_client,
+                billing_limiter,
+                token_dropper,
+                config.export_prometheus,
+                config.capture_mode,
+                config.concurrency_limit,
+                event_max_bytes,
+            )
+        } else {
+            router::router(
+                crate::time::SystemTime {},
+                liveness,
+                kafka_sink,
+                redis_client,
+                billing_limiter,
+                token_dropper,
+                config.export_prometheus,
+                config.capture_mode,
+                config.concurrency_limit,
+                event_max_bytes,
+            )
+        }
```
Feels a tad over-duplicated. Could we not assign the sink to a var and then just do the router:: stuff at the end?

frankh replied:
Yeah, I tried for ages to do that then gave up

Would need to have sink be type Box<dyn Events + Send + Sink> and then need to change router and everything to use boxed stuff and in the end felt less boilerplate to just duplicate the router creation

Definitely my least favourite part of rust the stupid known size stuff

frankh replied:
there you go: https://github.com/PostHog/posthog/pull/28042/commits/570d2fac55bd35cd5b812d571bf18ce25eac9088

benjackwhite reviewed (commented):
Generally I think this looks good but would definitely defer to a rusty given its capture.

I wonder if it would make sense to have an option which is "dual write" as well instead of fallback? That way we could run a mirror of say 0.01% of traffic with dual write enabled and be able to see live that it is writing the correct data

frankh committed:
refactor sink creation in server.rs to remove duplication

Box up the sink so that router is agnostic to whether its using a
KafkaSink or FallbackSink

frankh committed:
add periodic healthchecks to s3 bucket

previously we would only healthcheck on writing messages which is no
good if we're running as a fallback and expect long periods of time with
no messages being written

frankh committed:
Add healthstrategy options ALL/ANY for liveness checks

this gives us the option of being "live" if at least 1 component is up instead
of requiring all components be healthy

frankh committed:
update liveness immediately in kafka constructor and don't error on fail

we don't error if we fail to connect as there may be other sinks that report healthy

frankh committed:
only send to primary sink first if its healthy

frankh committed:
update rust version

frankh committed:
add s3 sink prefix config option

frankh committed:
only fallback on RetryableSinkError

frankh committed:
use buffer total_size not s3 response

the response only includes size if you're appending :/

frankh committed:
add logging on primary health changes in fallback sink

frankh committed:
fix rdkafka health check

stats.rx_bytes is a non-resetting counter, so is true if kafka has
ever been up, not if its currently up

frankh committed:
cargo fmt

frankh committed:
remove weird from_env

completely unecessary, this is what Envconfig does

frankh committed:
comments and slight refactors

frankh committed:
reduce log noise

oliverb123 commented on rust/capture/src/sinks/s3.rs:
```diff
@@ -0,0 +1,399 @@
+impl Drop for S3Sink {
+    fn drop(&mut self) {
+        if let Some(shutdown) = self.shutdown.take() {
+            // we don't actually need to send on the channel, dropping works
```
If you're not going to call `send`, this take and manual drop does nothing, and you can skip the manual drop impl altogether, but if I was you I'd do the `send` - it doesn't hurt.

frankh replied:
yeah i actually realised this and was gonna remove the drop - but yeah +1 for making explicit

oliverb123 commented on rust/capture/src/sinks/s3.rs:
```diff
@@ -0,0 +1,399 @@
+impl Event for S3Sink {
+    #[instrument(skip_all)]
+    async fn send(&self, event: ProcessedEvent) -> Result<(), CaptureError> {
+        let mut buffer = self.buffer.lock().await;
```
This buffer updating stuff, and the loop below, should be moved into an `impl EventBuffer` block, something like `pub fn add_event(&mut self, e: Event) -> Result<(), CaptureError>`. You should do the serialisation there, immediately rather than on flush, so that any single event/request can't poison the whole batch.

oliverb123 commented on rust/capture/src/sinks/s3.rs:
```diff
@@ -0,0 +1,399 @@
+            hostname
+        );
+
+        let mut data = BytesMut::new();
```
Generally, all this buffer fiddling could/should be moved into an `impl EventBuffer` block, with a `pub fn to_bytes(self) -> Vec<u8>`.

oliverb123 reviewed (commented):
Bunch of stuff here, some small and some more significant. Overall I like the approach, although if we're in a fallback state I'd like that to be REALLY LOUD.


If the `Fallback` sink exposed a prometheus `gauge` that's just set to 1 when it's in a fallback state, we could put an alert on that and fire a pagerduty alarm if it happens (if kafka's down we're incidenting anyway), and that'd make me feel a lot better. The counters you have in place for this would also work for the alarm - either is fine by me (or both, both is good), but I think a gauge would be marginally easier to add to a dashboard in a way that's obvious what the current state is.

frankh commented on rust/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM docker.io/lukemathwalker/cargo-chef:latest-rust-1.80.1-bookworm AS chef
+FROM docker.io/lukemathwalker/cargo-chef:latest-rust-1.82-bookworm AS chef
```
i bumped the top level dockerfile too - i think that's all?

frankh committed:
use workspace dependencies

frankh committed:
simplify flush error

frankh commented on rust/capture/src/sinks/s3.rs:
```diff
@@ -0,0 +1,399 @@
+
+                            let result = s3sink.flush_buffer(&mut old_buffer).await;
+                            if result.is_ok() {
+                                last_healthcheck = Instant::now();
```
the idea was supposed to make sure if we didn't send any events for 10 seconds the healthcheck wouldn't start failing - but yeah it got convoluted - i'll fix it 👍

frankh committed:
simplify writing to buffer

frankh committed:
cleanup flushing

frankh committed:
refactor S3Sink to use weak inner ref for thread logic

frankh committed:
correctly drop lock sooner

frankh committed:
don't unwrap()

can't really happen but better to ensure we never panic

frankh committed:
remove unneeded PartialEq

frankh committed:
nicer s3 object path

frankh committed:
add primary sink health gauge metric

frankh commented on rust/capture/src/config.rs:
```diff
@@ -61,6 +61,55 @@ pub struct Config {
+
+    #[envconfig(default = "false")]
+    pub s3_fallback_enabled: bool,
+    pub s3_fallback_bucket: Option<String>,
+    pub s3_fallback_endpoint: Option<String>,
```
that's to enable the feature, not whether its actively writing to the fallback or not
